File: Some_cpp_sols/C1553.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int Win(string s, int who){
    vector<int> goals(2,0);
    vector<int> r(2,5);
    for(int i=0;i<10;i++){
        if(goals[0] > goals[1] + r[1] || goals[1] > goals[0] + r[0])
            return i;
        else if(s[i] == '1')
            goals[i % 2]++;
        else if(s[i] == '?' && who == i % 2){
            goals[i % 2]++;
        }
        r[i % 2]--;
    }
    return 10;
}
int main(){
    int t;
    cin>>t;
    string s;
    while(t--){cin>>s;cout<<min(Win(s,0),Win(s,1))<<endl;}
    return 0;
}

File: C/C1553.java
import java.util.Scanner;

public class C1553 {
    public static int Win(char[] arr, int who){
        // Because two players are playing.
        int[] goals = new int[2];
        int[] r = new int[2];
        // It can be applied in general
        // As two players were playing so
        // Each got half and half
        r[0] = 5;r[1] = 5;
        for(int i=0;i<10;i++){
            // If the rounds are not enough to cover backlog;
            if(goals[0] >  goals[1] + r[1] || goals[1] > goals[0]  + r[0]){
                return i;
            }else if(arr[i] == '1'){ // Check if goal occurs
                goals[i % 2]++;
            }else if(arr[i] == '?' && who == i % 2){
                // Special case when "?" and we need who to win;
                goals[i % 2]++;
            }
            // Does not matter the thing is rounds are decreasing;
            r[i % 2]--;
        }
        return 10;
    }
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int t = input.nextInt();
        while(t--!=0){
            char[] arr = input.next().toCharArray();
            // Minimize round by watching players win
            // Hence get the minimal score
            System.out.println(Math.min(Win(arr,0),Win(arr, 1)));
        }
        input.close();
    }
}
